src/Orleans.Transactions/Abstractions/ITransactionFaultInjector.cs

```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Orleans.Transactions.Abstractions
+{
+    public interface ITransactionFaultInjector
+    {
+        void BeforeStore();
+        void AfterStore();
+    }
```

Comment:
Seems odd that a fault injector would be added into the public abstractions package... is there a better way?

Eg, could we generalize this to `ITransactionStorageObserver` or a name with a similarly general meaning?

Reply:
Even better, could this be implemented using something like a decorator pattern instead? It seems that we could decorate `ITransactionalStateStorage<TState>` as-needed and avoid special-casing the code. As a bonus, doing so would remove the class of errors whereby we forget to call the Before/After methods around the Store call.

Reply:
the storage exception injection logic is too deep into the implementation, if using decorator pattern, I might need to decorate every class.. I will see what I can do there. may decorate ITransactionalStateStorage is a good alternative. But then we need to decorate different storage provider implementation if we have multiple